in-progress/7588-spartan-clusters.md

```diff
@@ -0,0 +1,301 @@
+
+### Why not docker compose?
+
+We use docker-compose already in some tests. The problem is that it is very difficult to test a network with more than a few nodes. It is also difficult to simulate network conditions. Last, we wouldn't be able to use the same tooling to deploy a public network.
+
+The thinking here is to use the same tooling that we use for production deployments to test our networks. This should result in less code and more confidence.
+
+## Interface
+
+The users of this work are developers of aztec and possibly infrastructure providers.
+
+There will be a new top-level folder `helm-charts`. In it, there will be a `aztec-network` chart.
+
+A lot of the value is shown in the helm chart's values.yaml file. The values in this file are used to populate the templates that make up the rest of the chart. Here is an example:
+
+```yaml
+images:
+  aztec:
+    image: aztecprotocol/aztec:c07d910d0b8b83b62008e79a7085db6c5020df4e
+
+bootNode:
+  replicas: 1
+  service:
+    p2pPort: 40400
+    nodePort: 8080
+  logLevel: 'debug'
+  debug: 'discv5:*,aztec:*'
+  fakeProofs: false
+  resources:
+    requests:
+      cpu: 100m
+      memory: 128Gi
+    limits:
+      cpu: 100m
+      memory: 128Gi
+
+
+validator:
+  replicas: 32
+  service:
+    p2pPort: 40400
+    nodePort: 8080
+  logLevel: 'debug'
+  debug: 'discv5:*,aztec:*'
+  fakeProofs: false
+  resources: {}
+
+pxe:
+  replicas: 3
+  service:
+    type: ClusterIP
+    port: 8080
+    targetPort: 8080
+  fakeProofs: false
+  resources: {}
+
+chaos:
+  scenarios:
+    - name: 'boot-node-failure'
+      type: 'pod'
+      selector:
+        matchLabels:
+          app: boot-node
+      action: 'kill'
+      actionOptions:
+        force: true
+```
+
+The `yarn-projects/+end-to-end-base` earthly image will be updated to include `kubectl`, `helm`, and `kind`.
+
+We will add new tests to `yarn-projects/end-to-end`, and expose them to CI as earthly targets in the form `yarn-projects/end-to-end/+network-{test_name}`.
+
+The `ci` github action workflow will have a new job analogous to `e2e` and `bench-e2e` called `network-e2e`. This job will be a matrix on the `+network` targets, and for each it will ensure the test machine has a k8s cluster running with the relevant docker images loaded (via kind), and run a particular earthly target.
+
+Each target will:
+1. deploy a network to the k8s cluster (via helm install)
+2. (optional) apply some chaos configurations
+3. run helm tests
+4. tear down the network
+
+
+## Implementation
+
+### Helm Chart
+
+**anvil**
+
+There will be a deployment for anvil. It will have a single replica, and be exposed via ClusterIP.
+
+**boot node**
+
+There will be a statefulset with a single replica for the boot node. As part of its init container it will deploy the enshrined L1 contracts. Other nodes in the network will be able to resolve the boot node's address via its stable DNS name, e.g. `boot-node-0.aztec-network.svc.cluster.local`.
```

Comment:
I thought about taking the job route, but since nodes can just ask for addresses this seems simpler for now.

Good point on deploying L2 protocol contracts. I'll include that. 👍

Reply:
I think @just-mitch intends for the boot node to be a regular full node, not the P2P boot node.

But yeah, we will need to run a bootstrapping process similar to what we are currently doing on our other networks.